poundctl.8
.\" Pound - the reverse-proxy load-balancer
.\" Copyright (C) 2002-2010 Apsis GmbH
.\" Copyright (C) 2018-2022 Sergey Poznyakoff
.\"
.\" Pound is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Pound is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with pound.  If not, see <http://www.gnu.org/licenses/>.
.TH POUNDCTL 8 "November 19, 2022" "poundctl" "System Manager's Manual"
.SH NAME
poundctl \- control the pound daemon
.SH SYNOPSIS
.TP
.B poundctl \fI-c /path/to/socket\fR [\fI-L/-l\fR] [\fI-S/-s\fR] [\fI-B/-b\fR] [\fI-N/-n\fR] [\fI-H\fR] [\fI-X\fR]
.SH DESCRIPTION
.PP
.B Poundctl
controls various aspects of the operation of the
.I pound(8)
program.
.SH OPTIONS
Options available:
.TP
\fB\-c /path/to/socket\fR
The path to the (Unix-domain) socket
.B pound
was configured to listen on for control. Your
.B pound
configuration file must contain the directive
.I Control "/path/to/socket"
for
.B poundctl
to work.
.TP
\fB\-L/\-l \fIn\fR
Enable/disable a listener. A disabled listener will stop accepting connection
requests.
.TP
\fB\-S/\-s \fIn m\fR
Enable/disable a service. A disabled service will not be used by
.B pound
to answer requests.
.TP
\fB\-B/\-b \fIn m r\fR
Enable/disable a back-end. A disabled back-end will not be passed requests to
answer. Note however that existing sessions may still cause requests to be
sent their way.
.TP
\fB\-N \fIn m k r\fR
Add a session to service m in listener n. The session key is k and it points to
back-end r.
.TP
\fB\-n \fIn m k\fR
Remove a session from service m in listener n. The session key is k.
.PP
The parameters \fIn\fR, \fIm\fR and \fIr\fR refer to the number
assigned to a particular listener, service and back-end in the
listings. A listener number of -1 refers by convention to the global
context.
.TP
\fB\-H\fR
Try to resolve the addresses to symbolic names. Depending on your configuration,
this may require an active DNS.
.TP
\fB\-X\fR
Show the results in XML format.
.PP
When called without flags
.B poundctl
will print out a listing of the
.B pound
internal state.
.SH AUTHOR
Written by Robert Segall, Apsis GmbH, and Sergey Poznyakoff.
.SH "REPORTING BUGS"
Report bugs to <gray+pound@gnu.org.ua>.
.SH COPYRIGHT
Copyright \(co 2002-2010 Apsis GmbH.
.br
Copyright \(co 2018-2022 Sergey Poznyakoff
.sp
.na
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.sp
.ad
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.\" Local variables:
.\" eval: (add-hook 'write-file-hooks 'time-stamp)
.\" time-stamp-start: ".TH [A-Z_][A-Z0-9_.\\-]* [0-9] \""
.\" time-stamp-format: "%:B %:d, %:y"
.\" time-stamp-end: "\""
.\" time-stamp-line-limit: 20
.\" end:


NEWS
Pound -- history of user-visible changes. 2022-11-25
See the end of file for copying conditions.

Pound is a continuation of the software originally developed by
Robert Segall at Apsis GmbH, which was officially discontinued
on 2022-09-19.  See the README.md file for details.

Please send pound bug reports to <gray@gnu.org>

Version 4.0 (git)

* Support for OpenSSL 3.0

* Added testsuite.

* Fixes in configuration parsing.

=========================================================================
Copyright information:

Copyright (C) 2018-2022 Sergey Poznyakoff

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[  ]*$"
eval: (add-hook 'write-file-hooks 'time-stamp)
time-stamp-start: "changes. "
time-stamp-format: "%:y-%02m-%02d"
time-stamp-end: "\n"
end:




src/poundctl.c
/*
 * Pound - the reverse-proxy load-balancer
 * Copyright (C) 2002-2010 Apsis GmbH
 * Copyright (C) 2018-2022 Sergey Poznyakoff
 *
 * Pound is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pound is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pound.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contact information:
 * Apsis GmbH
 * P.O.Box
 * 8707 Uetikon am See
 * Switzerland
 * EMail: roseg@apsis.ch
 */
#include    "pound.h"

static int xml_out = 0;
static int host_names = 0;

static void
usage (const char *arg0)
{
  fprintf (stderr, "Usage: %s -c /control/socket [ -X ] cmd\n", arg0);
  fprintf (stderr, "\twhere cmd is one of:\n");
  fprintf (stderr, "\t-L n - enable listener n\n");
  fprintf (stderr, "\t-l n - disable listener n\n");
  fprintf (stderr,
	   "\t-S n m - enable service m in service n (use -1 for global services)\n");
  fprintf (stderr,
	   "\t-s n m - disable service m in service n (use -1 for global services)\n");
  fprintf (stderr,
	   "\t-B n m r - enable back-end r in service m in listener n\n");
  fprintf (stderr,
	   "\t-b n m r - disable back-end r in service m in listener n\n");
  fprintf (stderr,
	   "\t-N n m k r - add a session with key k and back-end r in service m in listener n\n");
  fprintf (stderr,
	   "\t-n n m k - remove a session with key k r in service m in listener n\n");
  fprintf (stderr, "\n");
  fprintf (stderr,
	   "\tentering the command without arguments lists the current configuration.\n");
  fprintf (stderr, "\tthe -X flag results in XML output.\n");
  fprintf (stderr,
	   "\tthe -H flag shows symbolic host names instead of addresses.\n");
  exit (1);
}

/*
 * Translate inet/inet6 address/port into a string
 */
static char *
prt_addr (const struct addrinfo *addr)
{
  static char res[UNIX_PATH_MAX + 9];
  char buf[UNIX_PATH_MAX];
  int port;
  void *src;

  memset (buf, 0, UNIX_PATH_MAX);
  switch (addr->ai_family)
    {
    case AF_INET:
      src = (void *) &((struct sockaddr_in *) addr->ai_addr)->sin_addr.s_addr;
      port = ntohs (((struct sockaddr_in *) addr->ai_addr)->sin_port);
      if (host_names
	  && !getnameinfo (addr->ai_addr, addr->ai_addrlen, buf,
			   UNIX_PATH_MAX - 1, NULL, 0, 0))
	break;
      if (inet_ntop (AF_INET, src, buf, sizeof (buf) - 1) == NULL)
	strncpy (buf, "(UNKNOWN)", sizeof (buf) - 1);
      break;

    case AF_INET6:
      src = (void *) &((struct sockaddr_in6 *) addr->ai_addr)->sin6_addr.s6_addr;
      port = ntohs (((struct sockaddr_in6 *) addr->ai_addr)->sin6_port);
      if (host_names
	  && !getnameinfo (addr->ai_addr, addr->ai_addrlen, buf,
			   sizeof (buf) - 1, NULL, 0, 0))
	break;
      if (inet_ntop (AF_INET6, src, buf, sizeof (buf) - 1) == NULL)
	strncpy (buf, "(UNKNOWN)", sizeof (buf) - 1);
      break;

    case AF_UNIX:
      strncpy (buf, (char *) addr->ai_addr, sizeof (buf) - 1);
      port = 0;
      break;

    default:
      strncpy (buf, "(UNKNOWN)", sizeof (buf) - 1);
      port = 0;
      break;
    }
  snprintf (res, sizeof (res) - 1, "%s:%d", buf, port);
  return res;
}

static int
be_prt (const int sock)
{
  BACKEND be;
  struct sockaddr_storage a, h;
  int n_be;
  ssize_t n;

  n_be = 0;
  while ((n = read (sock, (void *) &be, sizeof (BACKEND))) == sizeof (BACKEND))
    {
      if (be.disabled < 0)
	break;
      if (read (sock, &a, be.addr.ai_addrlen) == -1)
	{
	  perror ("read");
	  return -1;
	}
      be.addr.ai_addr = (struct sockaddr *) &a;
      if (be.ha_addr.ai_addrlen > 0)
	{
	  if (read (sock, &h, be.ha_addr.ai_addrlen) == -1)
	    {
	      perror ("read");
	      return -1;
	    }
	  be.ha_addr.ai_addr = (struct sockaddr *) &h;
	}
      if (xml_out)
	printf
	  ("<backend index=\"%d\" address=\"%s\" avg=\"%.3f\" priority=\"%d\" alive=\"%s\" status=\"%s\" />\n",
	   n_be++, prt_addr (&be.addr), be.t_average / 1000000, be.priority,
	   be.alive ? "yes" : "DEAD", be.disabled ? "DISABLED" : "active");
      else
	printf ("    %3d. Backend %s %s (%d %.3f sec) %s\n", n_be++,
		prt_addr (&be.addr), be.disabled ? "DISABLED" : "active",
		be.priority, be.t_average / 1000000,
		be.alive ? "alive" : "DEAD");
    }
  if (n == -1)
    {
      perror ("read");
      return -1;
    }
  return 0;
}

static int
sess_prt (const int sock)
{
  TABNODE sess;
  int n_be, n_sess, cont_len;
  char buf[KEY_SIZE + 1];
  ssize_t n;

  n_sess = 0;
  while ((n = read (sock, (void *) &sess, sizeof (TABNODE))) == sizeof (TABNODE))
    {
      if (sess.content == NULL)
	break;
      if (read (sock, &n_be, sizeof (n_be)) == -1 ||
	  read (sock, &cont_len, sizeof (cont_len)) == -1)
	{
	  perror ("read");
	  return -1;
	}

      memset (buf, 0, KEY_SIZE + 1);
      /* cont_len is at most KEY_SIZE */

      if (read (sock, buf, cont_len) == -1)
	{
	  perror ("read");
	  return -1;
	}

      if (xml_out)
	{
	  int i, j;
	  char escaped[KEY_SIZE * 2 + 1];

	  for (i = j = 0; buf[i]; i++)
	    if (buf[i] == '"')
	      {
		escaped[j++] = '\\';
		escaped[j++] = '"';
	      }
	    else
	      escaped[j++] = buf[i];
	  escaped[j] = '\0';
	  printf ("<session index=\"%d\" key=\"%s\" backend=\"%d\" />\n",
		  n_sess++, escaped, n_be);
	}
      else
	printf ("    %3d. Session %s -> %d\n", n_sess++, buf, n_be);
    }

  if (n == -1)
    {
      perror ("read");
      return -1;
    }
  return 0;
}

static int
svc_prt (const int sock)
{
  SERVICE svc;
  int n_svc;
  ssize_t n;
  int res = 0;

  n_svc = 0;
  while ((n = read (sock, (void *) &svc, sizeof (SERVICE))) == sizeof (SERVICE))
    {
      if (svc.disabled < 0)
	break;
      if (xml_out)
	{
	  if (svc.name[0])
	    printf ("<service index=\"%d\" name=\"%s\" status=\"%s\">\n",
		    n_svc++, svc.name, svc.disabled ? "DISABLED" : "active");
	  else
	    printf ("<service index=\"%d\"%s>\n", n_svc++,
		    svc.disabled ? " DISABLED" : "");
	}
      else
	{
	  if (svc.name[0])
	    printf ("  %3d. Service \"%s\" %s (%d)\n", n_svc++, svc.name,
		    svc.disabled ? "DISABLED" : "active", svc.tot_pri);
	  else
	    printf ("  %3d. Service %s (%d)\n", n_svc++,
		    svc.disabled ? "DISABLED" : "active", svc.tot_pri);
	}
      if ((res = be_prt (sock)) == 0)
	res = sess_prt (sock);
      if (xml_out)
	printf ("</service>\n");
      if (res)
	break;
    }

  if (n == -1)
    {
      perror ("read");
      res = -1;
    }
  return res;
}

static int
get_sock (const char *sock_name)
{
  struct sockaddr_un ctrl;
  int res;

  memset (&ctrl, 0, sizeof (ctrl));
  ctrl.sun_family = AF_UNIX;
  strncpy (ctrl.sun_path, sock_name, sizeof (ctrl.sun_path) - 1);
  if ((res = socket (PF_UNIX, SOCK_STREAM, 0)) < 0)
    {
      perror ("socket create");
      exit (1);
    }
  if (connect (res, (struct sockaddr *) &ctrl, (socklen_t) sizeof (ctrl)) < 0)
    {
      perror ("connect");
      exit (1);
    }
  return res;
}

int
main (const int argc, char **argv)
{
  CTRL_CMD cmd;
  int sock, n_lstn, i;
  char *arg0, *sock_name;
  int c_opt, en_lst, de_lst, en_svc, de_svc, en_be, de_be, a_sess, d_sess,
    is_set;
  LISTENER lstn;
  struct sockaddr_storage a;
  int status = 0;

  arg0 = *argv;
  sock_name = NULL;
  en_lst = de_lst = en_svc = de_svc = en_be = de_be = is_set = a_sess =
    d_sess = 0;
  memset (&cmd, 0, sizeof (cmd));
  opterr = 0;
  i = 0;
  while (!i && (c_opt = getopt (argc, argv, "c:LlSsBbNnXH")) > 0)
    switch (c_opt)
      {
      case 'c':
	sock_name = optarg;
	break;

      case 'X':
	xml_out = 1;
	break;

      case 'L':
	if (is_set)
	  usage (arg0);
	en_lst = is_set = 1;
	break;

      case 'l':
	if (is_set)
	  usage (arg0);
	de_lst = is_set = 1;
	break;

      case 'S':
	if (is_set)
	  usage (arg0);
	en_svc = is_set = 1;
	break;

      case 's':
	if (is_set)
	  usage (arg0);
	de_svc = is_set = 1;
	break;

      case 'B':
	if (is_set)
	  usage (arg0);
	en_be = is_set = 1;
	break;

      case 'b':
	if (is_set)
	  usage (arg0);
	de_be = is_set = 1;
	break;

      case 'N':
	if (is_set)
	  usage (arg0);
	a_sess = is_set = 1;
	break;

      case 'n':
	if (is_set)
	  usage (arg0);
	d_sess = is_set = 1;
	break;

      case 'H':
	host_names = 1;
	break;

      default:
	if (optopt == '1')
	  {
	    optind--;
	    i = 1;
	  }
	else
	  {
	    fprintf (stderr, "bad flag -%c", optopt);
	    usage (arg0);
	  }
	break;
      }

  if (sock_name == NULL)
    usage (arg0);
  if (en_lst || de_lst)
    {
      if (optind != (argc - 1))
	usage (arg0);
      cmd.cmd = (en_lst ? CTRL_EN_LSTN : CTRL_DE_LSTN);
      cmd.listener = atoi (argv[optind++]);
    }
  if (en_svc || de_svc)
    {
      if (optind != (argc - 2))
	usage (arg0);
      cmd.cmd = (en_svc ? CTRL_EN_SVC : CTRL_DE_SVC);
      cmd.listener = atoi (argv[optind++]);
      cmd.service = atoi (argv[optind++]);
    }
  if (en_be || de_be)
    {
      if (optind != (argc - 3))
	usage (arg0);
      cmd.cmd = (en_be ? CTRL_EN_BE : CTRL_DE_BE);
      cmd.listener = atoi (argv[optind++]);
      cmd.service = atoi (argv[optind++]);
      cmd.backend = atoi (argv[optind++]);
    }
  if (a_sess)
    {
      if (optind != (argc - 4))
	usage (arg0);
      cmd.cmd = CTRL_ADD_SESS;
      cmd.listener = atoi (argv[optind++]);
      cmd.service = atoi (argv[optind++]);
      memset (cmd.key, 0, KEY_SIZE + 1);
      strncpy (cmd.key, argv[optind++], KEY_SIZE);
      cmd.backend = atoi (argv[optind++]);
    }
  if (d_sess)
    {
      if (optind != (argc - 3))
	usage (arg0);
      cmd.cmd = CTRL_DEL_SESS;
      cmd.listener = atoi (argv[optind++]);
      cmd.service = atoi (argv[optind++]);
      strncpy (cmd.key, argv[optind++], KEY_SIZE);
    }
  if (!is_set)
    {
      if (optind != argc)
	usage (arg0);
      cmd.cmd = CTRL_LST;
    }

  sock = get_sock (sock_name);
  if (write (sock, &cmd, sizeof (cmd)) == -1)
    {
      perror ("write");
      exit (1);
    }

  if (!is_set)
    {
      int n;

      n_lstn = 0;
      if (xml_out)
	printf ("<pound>\n");
      if (read (sock, &n, sizeof (n)) == sizeof (n))
	{
	  if (xml_out)
	    printf ("<queue size=\"%d\"/>\n", n);
	  else
	    printf ("Requests in queue: %d\n", n);
	}
      while (read (sock, (void *) &lstn, sizeof (LISTENER)) == sizeof (LISTENER))
	{
	  if (lstn.disabled < 0)
	    break;
	  if (read (sock, &a, lstn.addr.ai_addrlen) == -1)
	    {
	      perror ("read");
	      break;
	    }
	  lstn.addr.ai_addr = (struct sockaddr *) &a;
	  if (xml_out)
	    printf
	      ("<listener index=\"%d\" protocol=\"%s\" address=\"%s\" status=\"%s\">\n",
	       n_lstn++,
	       !SLIST_EMPTY (&lstn.ctx_head) ? "HTTPS" : "http", prt_addr (&lstn.addr),
	       lstn.disabled ? "DISABLED" : "active");
	  else
	    printf ("%3d. %s Listener %s %s\n", n_lstn++,
		    !SLIST_EMPTY (&lstn.ctx_head) ? "HTTPS" : "http", prt_addr (&lstn.addr),
		    lstn.disabled ? "*D" : "a");
	  if (svc_prt (sock))
	    status = 1;
	  if (xml_out)
	    printf ("</listener>\n");
	}
      if (!xml_out)
	printf (" -1. Global services\n");
      if (svc_prt (sock))
	status = 1;
      if (xml_out)
	printf ("</pound>\n");
    }
  return status;
}
